### PR TITLE
[20933] Set DataSharing in Writer|ReaderProxyData

### PR DESCRIPTION
**Pull request by elianalf**
## Description

Previously, Writer|ReaderProxyData always reported DataSharing as AUTO. This PR set it to ON or OFF based on the PID_DATASHARING.

 @Mergifyio backport 2.14.x 2.13.x 2.10.x 2.6.x 

## Contributor Checklist

- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- N/A Any new configuration API has an equivalent XML API (with the corresponding XSD extension) 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- N/A New feature has been added to the `versions.md` file (if applicable).
- N/A New feature has been documented/Current behavior is correctly described in the documentation. 
    
- [x] Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Review by JesusPoderoso — APPROVED** [review]
LGTM with green CI

**Comment by elianalf** [comment]
@richiprosima please test_3 this

**Comment by elianalf** [comment]
@richiprosima please test_3 linux

**Review by JesusPoderoso — APPROVED** [review]
LGTM, good job!

File: test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp [hunk 1]
```diff
@@ -140,6 +140,90 @@ TEST(BuiltinDataSerializationTests, ok_with_defaults)
     }
 }
 
+TEST(BuiltinDataSerializationTests, msg_without_datasharing)
+{
+    {
+        // This was captured with wireshark from OpenDDS iShapes 3.16
```
Comment (EduPonz) [inline test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp:146]:
```suggestion
```

File: test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp [hunk 2]
```diff
@@ -140,6 +140,90 @@ TEST(BuiltinDataSerializationTests, ok_with_defaults)
...
+        msg.init(data_r_buffer, static_cast<uint32_t>(sizeof(data_r_buffer)));
+        msg.length = msg.max_size;
+
+        ReaderProxyData out(max_unicast_locators, max_multicast_locators);
+        out.readFromCDRMessage(&msg, network, false, true);
+        ASSERT_EQ(out.m_qos.data_sharing.kind(), OFF);
+    }
+
+    {
+        // This was captured with wireshark from OpenDDS iShapes 3.16
```
Comment (EduPonz) [inline test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp:163]:
```suggestion
```

File: test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp [hunk 3]
```diff
@@ -140,6 +140,90 @@ TEST(BuiltinDataSerializationTests, ok_with_defaults)
...
+        ReaderProxyData out(max_unicast_locators, max_multicast_locators);
+        out.readFromCDRMessage(&msg, network, false, true);
+        ASSERT_EQ(out.m_qos.data_sharing.kind(), OFF);
+    }
+}
+
+TEST(BuiltinDataSerializationTests, msg_with_datasharing)
+{
+    {
+        // This was captured with wireshark from OpenDDS iShapes 3.16
```
Comment (EduPonz) [inline test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp:184]:
```suggestion
```

File: test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp [hunk 4]
```diff
@@ -140,6 +140,90 @@ TEST(BuiltinDataSerializationTests, ok_with_defaults)
...
+        msg.init(data_r_buffer, static_cast<uint32_t>(sizeof(data_r_buffer)));
+        msg.length = msg.max_size;
+
+        ReaderProxyData out(max_unicast_locators, max_multicast_locators);
+        out.readFromCDRMessage(&msg, network, false, true);
+        ASSERT_EQ(out.m_qos.data_sharing.kind(), ON);
+    }
+
+    {
+        // This was captured with wireshark from OpenDDS iShapes 3.16
```
Comment (EduPonz) [inline test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp:205]:
```suggestion
```

**Review by EduPonz — CHANGES_REQUESTED** [review]

**Comment by EduPonz** [comment]
@Mergifyio backport 2.14.x 2.13.x 2.10.x 2.6.x

**Review by EduPonz — APPROVED** [review]

**Pull request merged**